bentoml/_internal/io_descriptors/numpy.py

```diff
@@ -546,15 +359,176 @@ async def predict(input: NDArray[np.int16]) -> NDArray[Any]:
             shape=sample_input.shape,
             enforce_dtype=enforce_dtype,
             enforce_shape=enforce_shape,
         )
-        inst.sample_input = sample_input
+        klass.sample_input = sample_input
+
+        return klass
 
-        return inst
+    async def from_grpc_request(
+        self, request: pb.Request, context: BentoServicerContext
+    ) -> ext.NpNDArray:
+        """
+        Process incoming protobuf request and convert it to ``numpy.ndarray``
+
+        Args:
+            request: Incoming RPC request message.
+            context: grpc.ServicerContext
+
+        Returns:
+            a ``numpy.ndarray`` object. This can then be used
+             inside users defined logics.
+        """
+        from ..utils.grpc import get_field
+        from ..utils.grpc import raise_grpc_exception
+        from ..utils.grpc import validate_content_type
+        from ..utils.grpc.mapping import dtypepb_to_fieldpb_map
+        from ..utils.grpc.mapping import dtypepb_to_npdtype_map
+        from ..utils.grpc.mapping import fieldpb_to_npdtype_map
+        from ..utils.grpc.mapping import npdtype_to_fieldpb_map
+
+        # validate gRPC content type if content type is specified
+        validate_content_type(context, self)
+
+        field = get_field(request, self)
+
+        # The logic behind processing shape is as follows:
+        # 1. If 'shape' is specified in the request message, then:
+        #    - if 'shape' is specified inside IO Descriptor, we will omit this value unless 'enforce_shape' is True.
+        #    - Otherwise, use 'shape' from message
+        #
+        # 2. If the 'shape' is not specified inside the request message, then:
+        #    - If 'shape' is not defined at the IO descriptor level, 'shape' value will be omitted unless 'enforce_shape' is True.
+        #    - Otherwise, we will use shape from the IO descriptor.
+        if self._enforce_shape:
+            logger.debug(
+                f"'enforce_shape={self._enforce_shape}', using '{self._shape}'..."
+            )
+            shape = self._shape
+        else:
+            shapepb = field.shape  # this can be an empty list ([])
+            if shapepb:
+                shape = tuple(shapepb)
+                if self._shape:
+                    logger.warning(
+                        f"'shape={self._shape}' is ignored when 'shape' is specified in request message. Using '{shape}' from request message..."
+                    )
+            else:
+                shape = self._shape
+
+        # The same logic for 'shape' applies to 'dtype' with the additional:
+        #
+        # 2a. If 'dtype' message is not specified, then we will infer given numpy dtype from the entry.
+        if self._enforce_dtype:
+            logger.debug(
+                f"'enforce_dtype={self._enforce_dtype}', using '{self._dtype}'..."
+            )
+            dtype = self._dtype
+            field_values = npdtype_to_fieldpb_map()[dtype]
```

Comment:
This is not the right behavior. let's just convert the numpy array from proto first, and then check with np.can_cast. reuse the validation function above.